test/e2e/federation-util.go

```diff
@@ -296,13 +296,20 @@ func cleanupServiceShardsAndProviderResources(namespace string, service *v1.Serv
 		err := wait.PollImmediate(framework.Poll, FederatedServiceTimeout, func() (bool, error) {
 			var err error
 			cSvc, err = c.Clientset.Services(namespace).Get(service.Name)
-			if err != nil && !errors.IsNotFound(err) {
-				// Get failed with an error, try again.
-				framework.Logf("Failed to find service %q in namespace %q, in cluster %q: %v. Trying again in %s", service.Name, namespace, name, err, framework.Poll)
+			if err == nil {
+				By(fmt.Sprintf("Service %q in namespace %q in cluster %q found", service.Name, namespace, name))
+				return true, nil
+			} else if errors.IsNotFound(err) {
+				// Not found yet try again.
+				framework.Logf("Service %q in namespace %q in cluster %q not found, trying again in %s: %v.", service.Name, namespace, name, framework.Poll, err)
+				cSvc = nil
 				return false, nil
+			} else {
+				// Get failed with some other error, fail.
+				framework.Logf("Failed to find service %q in namespace %q, in cluster %q: %v.", service.Name, namespace, name, err)
+				cSvc = nil
+				return false, err
```

Comment:
> Was the intention of the original code here to retry until the object exists, or to retry until we know whether or not the object exists?

It was the latter. Or rather, retry until we know the object really doesn't exist.

> The symptom I was seeing was that the resource it was trying to get and then delete did not exist, and it took some minutes to give up trying.

That's strange, because the previous logic had `&& !errors.IsNotFound(err)`. So if the object was not really found, the logic would cause the condition to `return true, nil` which would stop the retries.

> Presume we should implement the "ensureDeleted" logic. i.e. verify that the object doesn't exist, and if it does, then delete it? If so, yes the code in this PR is not correct, and I can send an update.

Right. But isn't this what the previous code did?